File: patterns/04-concurrency-patterns/02-worker-pool/python/worker_pool.py
class PatternError(Exception):
    def __init__(self, code):
        super().__init__(code)
        self.code = code

def evaluate(input_data):
    count = input_data["worker_count"]
    if count < 1:
        raise PatternError("invalid_worker_count")
    available = [0] * count
    used, executions = set(), []
    for job in input_data.get("jobs", []):
        if job["duration"] < 1:
            raise PatternError("invalid_duration")
        if job["outcome"] not in ("success", "failure"):
            raise PatternError("unknown_job_outcome")
        worker = min(range(count), key=lambda index: (available[index], index))
        start = available[worker]; finish = start + job["duration"]; available[worker] = finish; used.add(worker)
        executions.append({"id": job["id"], "worker": worker, "start": start, "finish": finish, "status": "completed" if job["outcome"] == "success" else "failed"})
    return {"executions": executions, "workers_used": len(used), "makespan": max(available, default=0), "joined": True}

File: patterns/04-concurrency-patterns/02-worker-pool/js/worker_pool.mjs
export class PatternError extends Error { constructor(code) { super(code); this.code = code; } }

export function evaluate(input) {
    if(input.worker_count<1)throw new PatternError("invalid_worker_count");const available=Array(input.worker_count).fill(0),used=new Set(),executions=[];
    for(const job of input.jobs??[]){if(job.duration<1)throw new PatternError("invalid_duration");if(!["success","failure"].includes(job.outcome))throw new PatternError("unknown_job_outcome");let worker=0;for(let index=1;index<available.length;index++)if(available[index]<available[worker])worker=index;const start=available[worker],finish=start+job.duration;available[worker]=finish;used.add(worker);executions.push({id:job.id,worker,start,finish,status:job.outcome==="success"?"completed":"failed"});}
    return{executions,workers_used:used.size,makespan:Math.max(0,...available),joined:true};
}
